hushline/premium.py

```diff
@@ -0,0 +1,445 @@
+
+
+def get_subscription(user: User) -> stripe.Subscription | None:
+    if user.stripe_subscription_id is None:
+        return None
+
+    return stripe.Subscription.retrieve(user.stripe_subscription_id)
+
+
+def handle_subscription_created(subscription: stripe.Subscription) -> None:
+    # Subscription is created, but not yet paid
+    user = db.session.query(User).filter_by(stripe_customer_id=subscription.customer).first()
+    if user:
+        user.stripe_subscription_id = subscription.id
+        db.session.commit()
+    else:
+        raise ValueError(f"Could not find user with customer ID {subscription.customer}")
+
+
+def handle_subscription_updated(subscription: stripe.Subscription) -> None:
+    # If subscription changes to cancel or unpaid, downgrade user
+    user = db.session.query(User).filter_by(stripe_subscription_id=subscription.id).first()
+    if user:
+        if subscription.status in ["canceled", "unpaid"]:
+            user.tier_id = FREE_TIER
+        db.session.commit()
+    else:
+        raise ValueError(f"Could not find user with subscription ID {subscription.id}")
+
+
+def handle_subscription_deleted(subscription: stripe.Subscription) -> None:
+    # If subscription is deleted, downgrade user
+    user = db.session.query(User).filter_by(stripe_subscription_id=subscription.id).first()
+    if user:
+        user.tier_id = FREE_TIER
+        user.stripe_subscription_id = None
+        db.session.commit()
+    else:
+        raise ValueError(f"Could not find user with subscription ID {subscription.id}")
+
+
+def handle_invoice_created(invoice: stripe.Invoice) -> None:
+    try:
+        new_invoice = StripeInvoice(invoice)
+        db.session.add(new_invoice)
+        db.session.commit()
+    except ValueError as e:
+        current_app.logger.error(f"Error creating invoice: {e}")
+
+
+def handle_invoice_payment_succeeded(invoice: stripe.Invoice) -> None:
+    stripe_invoice = db.session.query(StripeInvoice).filter_by(invoice_id=invoice.id).first()
+    if stripe_invoice:
+        stripe_invoice.amount_paid = invoice.amount_paid
+        stripe_invoice.amount_remaining = invoice.amount_remaining
+        stripe_invoice.status = StripeInvoiceStatusEnum(invoice.status)
+        db.session.commit()
+
+        # Invoice has been paid, so make sure the user is upgraded
+        user = db.session.query(User).filter_by(stripe_subscription_id=invoice.subscription).first()
+        if user:
+            user.tier_id = BUSINESS_TIER
+            db.session.commit()
+    else:
+        raise ValueError(f"Could not find invoice with ID {invoice.id}")
+
+
+async def worker(app: Flask) -> None:
+    current_app.logger.error("Starting worker")
+    with app.app_context():
+        while True:
+            stripe_event = (
+                db.session.query(StripeEvent)
+                .filter_by(status="pending")
+                .order_by(StripeEvent.created_at)
+                .first()
```

Comment:
I think this query should work. If there is a "pending" stripe event, it should update the status to "in_progress" and return the object, otherwise it returns `None`:
```suggestion
                db.session.scalars(db.update(StripeEvent)
                .where(status="pending")
                .order_by(StripeEvent.created_at)
                .values("in_progress")
                .limit(1)
                .returning(StripeEvent)).one_or_none()
```

Reply:
Fixed in https://github.com/scidsg/hushline/pull/595/commits/9c3af893d89787b5f716b42ddbe7849a8e1a276c